rdf-delta-server-local/src/main/java/org/seaborne/delta/zk/direct/ValidZooKeeperSupplier.java

```diff
@@ -0,0 +1,253 @@
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ *  See the NOTICE file distributed with this work for additional
+ *  information regarding copyright ownership.
+ */
+
+package org.seaborne.delta.zk.direct;
+
+import org.apache.zookeeper.*;
+import org.seaborne.delta.zk.ZkException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.function.Supplier;
+
+/**
+ * Supplies valid, connected {@link ZooKeeper} instances or throws an exception.
+ *
+ * <p>
+ * {@link ZooKeeper} instances will attempt to reconnect on their own if the connection is lost. However, session
+ * expiration causes an instance to become unusable. This class monitors connection state changes and will instantiate
+ * a fresh {@link ZooKeeper} instance when a session expires. This class also monitors the Ensemble configuration and
+ * will update the {@link #connectString} to reflect dynamic changes to the Ensemble.
+ * </p>
+ * <p>
+ * If all the Ensemble members present in the configuration leave the ensemble while the {@link ZooKeeper} instance
+ * is invalid, new connection attempts will fail resulting in a {@link ZkException} that will crash the application.
+ * This crash is essential to enabling the detection and replacement of a failed RDF Delta node.
+ * </p>
+ * <p>
+ * If {@link #close()} is called, calls to {@link #get()} will yield a {@link ZkException} that should crash the
+ * system. Only call {@link #close()} when you are done with this {@link ValidZooKeeperSupplier}.
+ * </p>
+ */
+public final class ValidZooKeeperSupplier implements Supplier<ZooKeeper>, Watcher, AutoCloseable {
+    /**
+     * {@link Logger}.
+     */
+    private static final Logger LOG = LoggerFactory.getLogger(ValidZooKeeperSupplier.class);
+
+    /**
+     * {@link Object} used for synchronizing asynchronous callbacks.
+     */
+    private final Object token = new Object();
+
+    /**
+     * The number of times to retry reconnect attempts before throwing a {@link ZkException}.
+     */
+    private final int retries;
+
+    /**
+     * A comma-separate list of host:port pairs pointing to {@link ZooKeeper} Ensemble members.
+     */
+    private volatile CharSequence connectString;
+
+    /**
+     * The current {@link ZooKeeper} instance.
+     */
+    private volatile ZooKeeper zooKeeper;
+
+    /**
+     * The current validity of the connection as determined by the connection monitor.
+     */
+    private volatile boolean isValid = false;
```

Comment:
The fourth state does exist when the connection is interrupted. Closing this object is how we tell it to stop trying to connect. Unfortunately, I'm caught in between vocabularies here in that `AutoClosable` has a `close` method and the connection state can be `CLOSED`, but they are not the same thing in the context of this class. As for using an enum, it would make more sense to me to skip listening for connection events and just check the state in the `get` method if you think that would be a preferable implementation.

Reply:
I pushed a revision. Please resolve this if you either approve or would rather move on.